### ISSUE TITLE
`Final[type[T]]` attribute breaks variance checking for generic type parameter `T`

### ISSUE THREAD

**Issue by uwu-420** [microsoft/pyright#11210]
**Describe the bug**
I think using `Final` is breaking the variance checks in my case. My example below stems from the intent to make a function expect a homogeneous list that shares a generic type with the other function parameter so it calling it with mismatching arguments will be caught by static type checking. Without the `Final` the check works and spots the error.

**Code or Screenshots**
```python
from typing import Final

class TypeWitness[T]:
    def __init__(self, t: type[T]):
        self.t: type[T] = t

class TypeWitnessFinal[T]:
    def __init__(self, t: type[T]):
        self.t: Final[type[T]] = t

def f[T](witness: TypeWitness[T], x: list[T]) -> None: ...

def f_final[T](witness: TypeWitnessFinal[T], x: list[T]) -> None: ...

# Gives this error as expected:
# Argument of type "list[int | str]" cannot be assigned to parameter "x" of type "list[T@f]" in function "f"
#    "Literal['foo']" is not assignable to "int"
f(TypeWitness(int), [42, "foo"])

# Unexpectedly, does not give an error.
f_final(TypeWitnessFinal(int), [42, "foo"])
```

**VS Code extension or command-line**
Pyright version 1.1.407 running both as VS Code extension and command-line tool.

**Comment by erictraut** [comment]
When you use `Final`, the field is considered read-only. Read-only fields are covariant whereas writable fields are invariant. That explains why `T` is inferred to be covariant in `TypeWitnessFinal`.

**Comment by uwu-420** [comment]
Thanks for the quick reply and the helpful explanation :)
